File: design_patterns/structural/decorator/python/src/decorator/logger/logger.py
from abc import ABC, abstractmethod
from datetime import datetime


class Logger(ABC):
    @abstractmethod
    def log(self, message: str) -> None:
        pass


class ConsoleLogger(Logger):
    def log(self, message: str) -> None:
        print(message)


class LoggerDecorator(Logger):
    def __init__(self, logger: Logger) -> None:
        self._logger = logger

    def log(self, message: str) -> None:
        self._logger.log(message)


class TimestampDecorator(LoggerDecorator):
    def log(self, message: str) -> None:
        return super().log(
            f"[{datetime.now().strftime('%Y-%m-%d %H:%M:%S')}] {message}"
        )


class LevelDecorator(LoggerDecorator):
    def __init__(self, logger: Logger, level: str) -> None:
        super().__init__(logger)
        self._level = level

    def log(self, message: str) -> None:
        return super().log(f"[{self._level}] {message}")


class FilenameDecorator(LoggerDecorator):
    def __init__(self, logger: Logger, filename: str) -> None:
        super().__init__(logger)
        self._filename = filename

    def log(self, message: str) -> None:
        return super().log(f"[{self._filename}] {message}")

File: design_patterns/structural/decorator/go/internal/logger/logger.go
package logger

import (
	"fmt"
	"time"
)

type Logger interface {
	Log(message string)
}

type ConsoleLogger struct{}

func (c *ConsoleLogger) Log(message string) {
	fmt.Printf("%s\n", message)
}

type LoggerDecorator struct {
	Logger Logger
}

func (d *LoggerDecorator) Log(message string) {
	d.Logger.Log(message)
}

type TimestampDecorator struct {
	LoggerDecorator
}

func (t *TimestampDecorator) Log(message string) {
	t.LoggerDecorator.Log(fmt.Sprintf("[%s] %s", time.Now().Format("2006-01-02 15:04:05"), message))
}

type LevelDecorator struct {
	LoggerDecorator
	Level string
}

func (l *LevelDecorator) Log(message string) {
	l.LoggerDecorator.Log(fmt.Sprintf("[%s] %s", l.Level, message))
}

type FileDecorator struct {
	LoggerDecorator
	Filename string
}

func (f *FileDecorator) Log(message string) {
	f.LoggerDecorator.Log(fmt.Sprintf("[%s] %s", f.Filename, message))
}
